### PR TITLE
Explore: Expand template variables when redirecting from dashboard panel

### LINKED ISSUE #19437: Explore: Template variables are not expanded in query when going from dashboard panel to Explore after reload
**What happened**:
ezgif com-video-to-gif (6)

When dashboard is created with template variables and you want to open it in Explore, you can open it:
1. With choosing Explore from panel menu. In this case, Grafana doesn't reload and request payload to database has expanded template variables -> that results in the correct data.
image.
However, query editor shows @hostname instead of specific expanded variables, which can be confusing.
image

2. With cmd+click on Explore option, which opens a new window. The request payload to database is in this case wrong (it uses @hostname as a searched field instead of specific fields)
image

**What you expected to happen**:
When I use cmd+click and new window open, it should send the query with expanded values. 
Moreover, when I click on Explore and the values are expanded in query, they should also be listed in the query editor, instead of seeing @hostname.

**How to reproduce it (as minimally and precisely as possible)**:
1. Open https://play.grafana.org/d/000000015/elasticsearch-templated?orgId=1
2. Select Explore from panel menu.
3. Cmd+click on Explore from panel menu.

**Environment**:
- Grafana version: 6.4.0-beat2

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR fixes expanding of template variables when redirecting from dashboard to explore by creating interpolateVariablesInQueries method for datasources.  
- [x] Prometheus
- [x] Graphite  
- [x] Elasticsearch  
- [x] Influxdb  
- [x] Postgres   
- [x] Loki    
- [x] Mssql    
- [x] Mysql

Issue for datasources below [has been created](https://github.com/grafana/grafana/issues/19602).
- [ ] Opentsdb
- [ ] Stackdriver
- [ ] Cloudwatch    
- [ ] Azure monitor             

**Which issue(s) this PR fixes**:
Fixes #19437

### CODE REVIEW

**Comment by ivanahuckova** [comment]
@davkal @aocenas @torkelo 
I have started to work on this issue and I have couple of questions/points for discussion. For all datasources, that don't have `getExploreState` method, [template variables are not expanded when redirecting to Explore in new window](https://github.com/grafana/grafana/blob/master/public/app/core/utils/explore.ts#L82). Also, there is a lot of datasources, that don't have this method (actually, only Prometheus and Graphite (where I've just added it). So my question is - if we want to add `getExploreState` to all of our datasources. Or if we want to implement it in [getExploreUrl](https://github.com/grafana/grafana/blob/master/public/app/core/utils/explore.ts#L53) and require templateSrv as a parameter, so we could use replace method. Or if we should do combination. 
Or maybe you have some different idea?

File: public/app/plugins/datasource/graphite/datasource.ts [hunk 1]
```diff
@@ -117,6 +121,26 @@ export class GraphiteDatasource {
     return tags;
   }
 
+  getExploreState(queries: GraphiteQuery[]): Partial<ExploreUrlState> {
```
Comment (torkelo) [inline public/app/plugins/datasource/graphite/datasource.ts]: I think we need to change this function to not be so Explore specific or have to use explore types (that are not exposed to plugins as they are internal only). 

This should not need to know about ExploreUrlState, think this function should only be:
interpolateVariablesInQueries(queries: DataQuery[]): DataQuery[]  

Will make it easier for data sources to implement and they do not need to use explore specific types.

Comment (ivanahuckova) [reply public/app/plugins/datasource/graphite/datasource.ts]: I agree. So does this mean that we should create `interpolateVariablesInQueries` function that needs to be implemented in every datasource? Also then for plugin datasources, do we want to add in documentation, that they should implement `interpolateVariablesInQueries` function?

File: public/app/plugins/datasource/graphite/datasource.ts [hunk 2]
```diff
@@ -117,6 +120,20 @@ export class GraphiteDatasource {
...
+        const expandedQuery = {
+          ...query,
+          target: this.templateSrv.replace(query.target),
+        };
+        return expandedQuery;
+      });
+    }
+    return expandedQueries;
+  }
+
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/graphite/datasource.ts]: @torkelo does this look good for function, that basically just interpolates variables and then it is used in explore. 🙂

File: public/app/plugins/datasource/graphite/datasource.ts (hunk 1 above)
Comment (torkelo) [reply public/app/plugins/datasource/graphite/datasource.ts]: it needs to be optional

File: public/app/plugins/datasource/graphite/datasource.ts (hunk 2 above)
Comment (torkelo) [reply public/app/plugins/datasource/graphite/datasource.ts]: yes, looks ok!

File: public/app/plugins/datasource/influxdb/datasource.ts [hunk 3]
```diff
@@ -176,6 +176,28 @@ export default class InfluxDatasource extends DataSourceApi<InfluxQuery, InfluxO
     return false;
   }
 
+  interpolateVariablesInQueries(queries: InfluxQuery[]): InfluxQuery[] {
+    let expandedQueries = queries;
+    if (queries && queries.length > 0) {
+      expandedQueries = queries.map(query => {
+        const expandedTags = query.tags.map(tag => {
```
Comment (torkelo) [inline public/app/plugins/datasource/influxdb/datasource.ts]: think this needs to check for raw queries as well (that only have the text field)

File: public/app/plugins/datasource/influxdb/datasource.ts [hunk 4]
```diff
@@ -176,6 +176,32 @@ export default class InfluxDatasource extends DataSourceApi<InfluxQuery, InfluxO
...
+        });
+        const expandedQuery = {
+          ...query,
+          datasource: this.name,
+          tags: expandedTags,
+          measurement: this.templateSrv.replace(query.measurement, null, 'regex'),
+        };
+        if (query.rawQuery) {
+          expandedQuery.query = this.templateSrv.replace(query.query, null, 'regex');
+        }
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/influxdb/datasource.ts]: @torkelo fix of the **check for raw queries** is here.

File: public/app/plugins/datasource/influxdb/datasource.ts [hunk 5]
```diff
@@ -176,6 +176,32 @@ export default class InfluxDatasource extends DataSourceApi<InfluxQuery, InfluxO
     return false;
   }
 
+  interpolateVariablesInQueries(queries: InfluxQuery[]): InfluxQuery[] {
+    let expandedQueries = queries;
+    if (queries && queries.length > 0) {
```
Comment (torkelo) [inline public/app/plugins/datasource/influxdb/datasource.ts]: if no queries return early [] (reduces nesting in rest of function).

File: public/app/plugins/datasource/influxdb/datasource.ts (hunk 3 above)
Comment (torkelo) [reply public/app/plugins/datasource/influxdb/datasource.ts]: If it's a rawQuery then there is no 100& guarantee that tags property exists (could be undefined)

**Review by torkelo — APPROVED** [review]
Looks good, just one minor thing

File: public/app/plugins/datasource/influxdb/datasource.ts (hunk 5 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/influxdb/datasource.ts]: ✅

**Pull request merged**